### PR TITLE
Rakri/cosine bug fix

### PR DESCRIPTION
**Pull request by rakri**
- [ ] Does this PR have a descriptive title that could go in our release notes?
- [ ] Does this PR add any new dependencies?
- [ ] Does this PR modify any existing APIs?
   - [ ] Is the change to the API backwards compatible?
- [ ] Should this result in any changes to our documentation, either updating existing docs or adding new ones?
 
#### What does this implement/fix? Briefly explain your changes.
Earlier build_disk_index did not allow cosine metric but search_disk_index did, which led to erroneous/undefined behaviour. now we fixed that, and fixed some associated bugs.

### CODE REVIEW

File: src/disk_utils.cpp [hunk 1]
```diff
@@ -1188,12 +1193,24 @@ int build_disk_index(const char *dataFilePath, const char *indexFilePath, const
...
         diskann::save_bin<float>(norm_file, &max_norm_of_base, 1, 1);
         diskann::cout << timer.elapsed_seconds_for_step("preprocessing data for inner product") << std::endl;
+        created_temp_file_for_processed_data = true;
+    }
+    else if (compareMetric == diskann::Metric::COSINE)
+    {
+        Timer timer;
+        std::cout << "Normalizing data for cosine to temporary file, please ensure there is additional "
+                     "(n*d*4) bytes for storing normalized base vectors, "
+                     "apart from the intermin indices and final index."
```
Comment (daxpryce) [inline src/disk_utils.cpp:1208]: what does "intermin indices" mean? intermediate?

Comment (rakri) [reply src/disk_utils.cpp]: sorry i meant intermediate indices, i.e., the in-mem indices created by DiskANN before the final disk index is created.
Fixed this, thanks for pointing out!

**Review by daxpryce — APPROVED** [review]
As far as I can tell this looks good

**Review by harsha-simhadri — CHANGES_REQUESTED** [review]
@rakri Add a CI test for cosine case with un-normalized input data?

**Comment by rakri** [comment]
> @rakri Add a CI test for cosine case with un-normalized input data?

@harshasi Added.

File: src/disk_utils.cpp [hunk 2]
```diff
@@ -1195,6 +1196,10 @@ int build_disk_index(const char *dataFilePath, const char *indexFilePath, const
     std::string disk_pq_pivots_path = index_prefix_path + "_disk.index_pq_pivots.bin";
     // optional, used if disk index must store pq data
     std::string disk_pq_compressed_vectors_path = index_prefix_path + "_disk.index_pq_compressed.bin";
+    std::string prepped_base =
+        index_prefix_path +
+        "_prepped_base.bin"; // temp file for storing pre-processed base file for cosine/ mips metrics
```
Comment (gopalrs) [inline src/disk_utils.cpp:1201]: Maybe we can extract the preprocessing code into a function like we did for DLVS? https://msasg.visualstudio.com/Bing_and_IPG/_git/ObjectStore?path=/private/Pegasus/src/DssmSearch/randnsg/src/aux_utils.cpp

File: src/pq_flash_index.cpp [hunk 3]
```diff
@@ -32,14 +32,16 @@ template <typename T, typename LabelT>
...
     : reader(fileReader), metric(m), _thread_data(nullptr)
 {
+    diskann::Metric metric_to_invoke = m;
     if (m == diskann::Metric::COSINE || m == diskann::Metric::INNER_PRODUCT)
     {
         if (std::is_floating_point<T>::value)
         {
-            diskann::cout << "Cosine metric chosen for (normalized) float data."
-                             "Changing distance to L2 to boost accuracy."
+            diskann::cout << "Since data is floating point, we assume that it has been appropriately pre-processed "
```
Comment (gopalrs) [inline src/pq_flash_index.cpp:40]: This looks ok, but can you please make sure it is in sync with the DLVS code?

**Review by gopalrs — APPROVED** [review]

File: src/disk_utils.cpp (hunk 2 above)
Comment (rakri) [reply src/disk_utils.cpp:1201]: Let us do this clean-up in another PR..

**Pull request merged**